### PR TITLE
Enable megacore for splash attention dkv backward

### PR DESCRIPTION
**Pull request by huytransformer**
# Description

This PR adds the flag `sa_bwd_dkv_megacore`. When set to True, the static dkv grid changes from
(kv_steps, num_q_heads, q_steps) to (num_kv_heads, kv_steps, q_heads_per_kv_head, q_steps) and the leading dimension becomes "parallel". dk/dv accumulate within one KV head group, so groups are independent and can run on separate cores. With dq_reduction_steps=3 the flag forces the static dkv grid. 

MQA and num_kv_heads=1 raise a ValueError.

## Performance

v5p-128, llama3.1-8b with head_dim=256, 16 query heads, 4 KV heads, ring CP=64, block sizes 1024, synthetic data:

| Config | Context | MFU flag off | MFU flag on | Step time |
|---|---|---|---|---|
| batch 1, dq_reduction_steps=0 | 256K | 34.5% [Xprof](http://xprof.corp.google.com/trace_viewer/htn-11683570770883506035) | 46.7% [Xprof](http://xprof.corp.google.com/trace_viewer/htn-17415528724194145841) | -26% |
| batch 1, dq_reduction_steps=0 | 1M | 39.0% [Xprof](http://xprof.corp.google.com/trace_viewer/htn-17601052804012555500) | 57.7% [Xprof](http://xprof.corp.google.com/trace_viewer/htn-4772183944023096872) | -32% |
| batch 4, dq_reduction_steps=0 | 256K | | | -0.4% |
| batch 1, dq_reduction_steps=3 | 256K | 37.6% | 51.2% | -27% |
| batch 1, dq_reduction_steps=3 | 1M | 43.0% | 61.4% | -30% |

 The flag is observed to be only useful at per-device batch 1.

Losses identical to the baseline at every step and same peak memory

# Tests

- `python3 -m pytest tests/unit/kernels_test.py -k Megacore`: dq/dk/dv are bitwise identical with the flag on and off for GQA and MHA in interpret mode on CPU. 
- The TPU runs in the table, 8 steps each on v5p-128

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 1]
```diff
@@ -1200,16 +1203,24 @@ def _flash_attention_dkv_kernel(
...
+      )
     grid_idx = (kv_index * q_steps) + q_index
     should_initialize = q_index == 0
     should_write = True if q_steps <= 2 else q_index == q_steps - 1
     if q_heads_per_kv_head > 1:
-      q_head_index_per_kv_head = lax.rem(q_head, q_heads_per_kv_head)
+      if not megacore_groups:
+        q_head_index_per_kv_head = lax.rem(q_head, q_heads_per_kv_head)
       should_initialize = jnp.logical_and(should_initialize, q_head_index_per_kv_head == 0)
       should_write = jnp.logical_and(should_write, q_head_index_per_kv_head == q_heads_per_kv_head - 1)
```
[the remark this answers is not in the document]
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1225]: The heads of a group accumulate dk/dv into a scratch buffer, and the last head writes out the total. This is the same path the default config uses today.

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -733,6 +733,7 @@ class SplashAttention(BaseModel):
   sa_block_q_dq: int = Field(512, description="Block size for Q_dq in splash attention.")
   sa_block_kv_dq: int = Field(512, description="Block size for KV_dq in splash attention.")
   sa_use_fused_bwd_kernel: bool = Field(False, description="Use fused backward kernel in splash attention.")
+  sa_bwd_dkv_megacore: bool = Field(False, description="Megacore-parallel kv-head groups in the static dkv grid.")
   sa_q_layout: str = Field("HEAD_DIM_MINOR", description="Layout for Q in splash attention.")
```
[the remark this answers is not in the document]
Comment (huytransformer) [reply src/maxtext/configs/types.py:740]: Agreed and fixed.

File: tests/unit/kernels_test.py [hunk 3]
```diff
@@ -179,5 +182,47 @@ def test_ragged_gqa_cpu(self):
...
+    reference = self._grads(bwd_dkv_megacore=False, num_kv_heads=self.num_query_heads)
+    megacore = self._grads(bwd_dkv_megacore=True, num_kv_heads=self.num_query_heads)
+    for expected, actual in zip(reference, megacore):
+      np.testing.assert_array_equal(expected, actual)
+
+  def test_megacore_rejects_single_kv_head(self):
+    with self.assertRaisesRegex(ValueError, "more than one KV head"):
+      self._grads(bwd_dkv_megacore=True, num_kv_heads=1)
+
+
```
[the remark this answers is not in the document]
Comment (huytransformer) [reply tests/unit/kernels_test.py:248]: Added

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 4]
```diff
@@ -1200,16 +1203,24 @@ def _flash_attention_dkv_kernel(
...
+          pl.program_id(1),
+          pl.program_id(2),
+          pl.program_id(3),
+      )
+    else:
+      kv_index, q_head, q_index = (
+          pl.program_id(0),
+          pl.program_id(1),
+          pl.program_id(2),
+      )
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1217]: I expect the pl.BlockSpec being different due to this change, but it seems not the case?

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 5]
```diff
@@ -1426,6 +1443,16 @@ def mask_index_map(h, grid_idx, rows_ref, cols_ref, mask_next_ref=None, *_):
       next_m = to_i32(mask_next_ref[grid_idx])  # pyrefly: ignore[unsupported-operation]
       return next_m, 0, 0
 
+  elif megacore_groups:
+    unravel = lambda f: lambda g, j, gi, i, *_: f(g * q_heads_per_kv_head + gi, i, j)
+    grid = (num_kv_heads, kv_steps, q_heads_per_kv_head, q_steps)
+
+    def mask_index_map(g, j, gi, i, rows_ref, cols_ref, mask_next_ref=None, *_):
+      del g, gi, rows_ref, cols_ref  # Unused.
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1451]: I don't think del works in jax.jit

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 6]
```diff
@@ -1749,7 +1777,13 @@ def _bwd_cost_estimate(
         # 2) for kv_seq_len, the splash attention prefetch schedule assumes no
         #     megacore
         # 3) for q_seq_len, we are reducing over it to compute dkv
-        compiler_params=pltpu.CompilerParams(dimension_semantics=("arbitrary",) * len(grid)),
+        # With megacore_groups the kv-head group dimension is independent (each
+        # group owns its dk/dv/dq blocks), so it is marked parallel.
+        compiler_params=pltpu.CompilerParams(
+            dimension_semantics=(("parallel",) + ("arbitrary",) * (len(grid) - 1))
+            if megacore_groups
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1784]: do you think perf improvement using parallel?

File: tests/unit/kernels_test.py [hunk 7]
```diff
@@ -179,5 +182,62 @@ def test_ragged_gqa_cpu(self):
     )
 
 
+class SplashAttentionDkvMegacoreCpuTest(unittest.TestCase):
+  """Tests for the splash attention dkv backward kv-head group split on CPU (interpret mode)."""
```
Comment (NuojCheng) [inline tests/unit/kernels_test.py:186]: The correctness test (with computation) on CPU?

File: tests/unit/tokamax_ring_attention_test.py [hunk 8]
```diff
@@ -181,6 +182,7 @@ def test_build_splash_config_disables_staged_dq_for_small_kv_shards(self):
         use_splash_scheduler=False,
         ring_scan_unroll=1,
         context_parallel_load_balance=False,
+        sa_bwd_dkv_megacore=False,
     )
```
Comment (NuojCheng) [inline tests/unit/tokamax_ring_attention_test.py:186]: please add a TPU test with new feature being true

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py (hunk 5 above)
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1451]: Ya del only removes the unused python variable name. other two mask_index_maps in this file do the same thing (lines 1442 and 1461)

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py (hunk 4 above)
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1217]: line 1447 maps the new grid indices back to the old head index (h = g * q_heads_per_kv_head + gi) so it's not different

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py (hunk 6 above)
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1784]: yes

File: tests/unit/tokamax_ring_attention_test.py (hunk 8 above)
Comment (huytransformer) [reply tests/unit/tokamax_ring_attention_test.py:186]: added

File: tests/unit/kernels_test.py (hunk 7 above)
Comment (huytransformer) [reply tests/unit/kernels_test.py:186]: yes via Interpret=True

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 9]
```diff
@@ -1200,16 +1203,24 @@ def _flash_attention_dkv_kernel(
...
-    kv_index, q_head, q_index = (
-        pl.program_id(0),
-        pl.program_id(1),
-        pl.program_id(2),
-    )
+    if megacore_groups:
+      kv_index, q_head_index_per_kv_head, q_index = (
+          pl.program_id(1),
+          pl.program_id(2),
+          pl.program_id(3),
```
Comment (dandragona) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1210]: Let's just add a comment that these indices shifted down since we now have a 4D grid

**Comment by dandragona** [comment]
I believe this change would also warrant a new tile size block sweep as the pipelining should be changed with the flag enabled.

**Review by dandragona — APPROVED** [review]

**Review by NuojCheng — APPROVED** [review]
